lib/web/join_tokens.go

```diff
@@ -239,20 +240,41 @@ func createJoinToken(ctx context.Context, m nodeAPIGetter, roles types.SystemRol
 	}, nil
 }
 
-func getJoinScript(settings scriptSettings, m nodeAPIGetter) (string, error) {
-	// Skip decoding validation for IAM tokens since they are generated with a different method
-	if settings.joinMethod != string(types.JoinMethodIAM) {
-		// This token does not need to be validated against the backend because it's not used to
-		// reveal any sensitive information. However, we still need to perform a simple input
-		// validation check by verifying that the token was auto-generated.
-		// Auto-generated tokens must be encoded and must have an expected length.
-		decodedToken, err := hex.DecodeString(settings.token)
+func getJoinScript(ctx context.Context, settings scriptSettings, m nodeAPIGetter) (string, error) {
+	switch settings.joinMethod {
+	case string(types.JoinMethodUnspecified), string(types.JoinMethodToken), string(types.JoinMethodIAM):
+		if settings.joinMethod != string(types.JoinMethodIAM) {
+			decodedToken, err := hex.DecodeString(settings.token)
+			if err != nil {
+				return "", trace.Wrap(err)
+			}
+			if len(decodedToken) != auth.TokenLenBytes {
+				return "", trace.BadParameter("invalid token %q", decodedToken)
+			}
+		}
+
+		// The provided token can be attacker controlled, so we must validate
+		// it with the backend before using it to generate the script.
+		_, err := m.GetToken(ctx, settings.token)
 		if err != nil {
-			return "", trace.Wrap(err)
+			return "", trace.BadParameter("invalid token")
 		}
+	default:
+		return "", trace.BadParameter("join method %q is not supported via script", settings.joinMethod)
+	}
 
-		if len(decodedToken) != auth.TokenLenBytes {
-			return "", trace.BadParameter("invalid token length")
+	// We must also validate the label spec, which can be controlled by
+	// an attacker and is fed into the join script.
+	labels, err := client.ParseLabelSpec(settings.nodeLabels)
+	if err != nil {
+		return "", trace.BadParameter("invalid node-labels")
+	}
+	for k, v := range labels {
+		if !types.IsValidLabelKey(k) {
+			return "", trace.BadParameter("invalid label key %q", k)
+		}
+		if strings.ContainsAny(v, "\r\n'#") {
```

Comment:
I think with the single quotes in the bash script now we should be okay for these, but to play it safe we could block these too
```suggestion
		if strings.ContainsAny(v, "\r\n'#`\\") {
```